### ISSUE TITLE
"Identifier expected" errors

### ISSUE THREAD

**Issue by GitMensch** [LADSoft/OrangeC#456]
I'm in the process of building MPFR (adjusting libtool behavior again and again is kind of getting boring, but I always get somewhere new...). As this is off-topic I'll outline the process below (part of it also applies to MPIR/GMP and likely to GnuCOBOL built on the command line).

The errors show when building [tests/tfits.c][2]:

~~~
$ occ --nologo -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DLT_OBJDIR=\".libs/\" -DHAVE_LITTLE_ENDIAN=1 -DTIME_WITH_SYS_TIME=1 -DHAVE_LOCALE_H=1 -DHAVE_WCHAR_H=1 -DHAVE_STDARG=1 -DHAVE_SYS_TIME_H=1 -DHAVE_STRUCT_LCONV_DECIMAL_POINT=1 -DHAVE_STRUCT_LCONV_THOUSANDS_SEP=1 -DHAVE_ALLOCA_H=1 -DHAVE_VA_COPY=1 -DHAVE_SETLOCALE=1 -DHAVE_GETTIMEOFDAY=1 -DHAVE_SIGNAL=1 -DHAVE_SIGACTION=1 -DHAVE_LONG_LONG=1 -DHAVE_INTMAX_T=1 -DMPFR_HAVE_INTMAX_MAX=1 -DMPFR_PRINTF_MAXLM=\"j\" -DMPFR_HAVE_NORETURN=1 -DMPFR_HAVE_FESETROUND=1 -DHAVE_DENORMS=1 -DHAVE_DENORMS_FLT=1 -DMPFR_NANISNAN=1 -DHAVE_ROUND=1 -DHAVE_TRUNC=1 -DHAVE_FLOOR=1 -DHAVE_CEIL=1 -DHAVE_NEARBYINT=1 -DMPFR_USE_THREAD_SAFE=1 -DMPFR_USE_C11_THREAD_SAFE=1 -DMPFR_USE_STATIC_ASSERT=1 -DNPRINTF_J=1 -DPRINTF_L=1 -DNPRINTF_T=1 -DMPFR_LONG_WITHIN_LIMB=1 -I. -I../../tests  -DSRCDIR='"../../tests"' -I../../src -I../src -I/d/Programme/Entwicklung/mpfr-4.0.2/build_occ/external -D__GMP_LIBGMP_DLL  -pa--gas -c -o tfits.o ../../tests/tfits.c
Error( 16)    ../../tests/tfits.c(141):  Identifier expected in function main
Error( 16)    ../../tests/tfits.c(145):  Identifier expected in function main
Error( 16)    ../../tests/tfits.c(149):  Identifier expected in function main
Error( 16)    ../../tests/tfits.c(153):  Identifier expected in function main
Error( 16)    ../../tests/tfits.c(157):  Identifier expected in function main
Error( 16)    ../../tests/tfits.c(161):  Identifier expected in function main
Error( 16)    ../../tests/tfits.c(165):  Identifier expected in function main
Error( 16)    ../../tests/tfits.c(170):  Identifier expected in function main
Error( 16)    ../../tests/tfits.c(172):  Identifier expected in function main
Error( 16)    ../../tests/tfits.c(176):  Identifier expected in function main
Error( 16)    ../../tests/tfits.c(179):  Identifier expected in function main
Error(114)    ../../tests/tfits.c(179):  Type mismatch in parameter '1' in call to 'flags_out' in function main
Error( 18)    ../../tests/tfits.c(179):  Undefined symbol 'e' in function main
Error(114)    ../../tests/tfits.c(179):  Type mismatch in parameter '1' in call to 'flags_out' in function main
Error( 18)    ../../tests/tfits.c(179):  Undefined symbol 'x'
Error( 18)    ../../tests/tfits.c(179):  Undefined symbol '_mpfr_exp'
Error(437)    ../../tests/tfits.c(180):  Invalid use of type 'int'
Error(437)    ../../tests/tfits.c(181):  Invalid use of type 'int'
Error(437)    ../../tests/tfits.c(182):  Invalid use of type 'int'
Error(437)    ../../tests/tfits.c(183):  Invalid use of type 'int'
Error(437)    ../../tests/tfits.c(184):  Invalid use of type 'int'
Error(437)    ../../tests/tfits.c(185):  Invalid use of type 'int'
Error(437)    ../../tests/tfits.c(186):  Invalid use of type 'int'
Error(437)    ../../tests/tfits.c(187):  Invalid use of type 'int'
Error(437)    ../../tests/tfits.c(190):  Invalid use of type 'int'
Error(  1)    ../../tests/tfits.c(190):  Too many errors or warnings
~~~

I'm a little lost here because of the huge amount of preprocessor macros using preprocessor macros using...

@LADSoft Do you think you could have a look at this?

Note: all other files [with the adjustments mentioned above] of both the library and test programs compile+link :-)

------------------------
Building MPFR is more or less the same as GMP (omitting yasm, adding gmp options during configure including `CPPFLAGS=-D__GMP_LIBGMP_DLL`).
In any case you need to have GMP (matching header, library, dll) I've fetched them from [the old GnuCOBOL prerequisite for VC](https://sourceforge.net/projects/open-cobol/files/win_prerequisites/) ([adjusting gmp.h to have the dllexport stuff defined for OrangeC](https://github.com/wbhart/mpir/commit/ff9d361633f323fb72bb8abf225bdb851a7012a2#diff-39863e45af7987e9ede3a9162b073d49)) and created the import library with `oimplib -C mpir.dll gmp.l`.
To separate GMP from the rest I've did the following: create directory "external", placed mpir.dll+gmp.l+gmp.h there and then configured the stuff (without omake/ogrep to not extend the test to these tools yet):

```sh
$ export PATH=$(PWD)/external:/path/to/Orangec/bin:$PATH
$ ../configure 'CC=occ --nologo' 'CFLAGS=-pa--gas' LD='olink --nologo' 'AR=olib --nologo' 'CPP=ocpp --nologo' CPP_FOR_BUILD='ocpp --nologo' 'CXX=occ --nologo' 'CXXCPP=ocpp --nologo' 'NM=onm --nologo' 'CPPFLAGS=-IX:\\path\\to\\external -D__GMP_LIBGMP_DLL' 'LDFLAGS=-LX:\\path\\to\\external'
```
Then manually adjust libtool/config.status (changing `lib -OUT:` to `olib --nologo -o`), then [adjust ltmain.sh][1], then ran and recognizing something is missing: change config.status to remove `--output-def` [the place where `--export-all-symbols` would also be removed] and ``func_echo_all "$deplibs" | $SED '\''s/ -lc$//'\''` -link -dll`; also, because of #453, `-DHAVE_FPC_CSR=1` , then go on with `make` (config.status will be auto-executed).

There were some linking issues which I've worked around by a manual link, which also provided me with a shared library libmpfr-6.dll + libmpfr-6.l. Then adjusted all libmpfr.la[i] to read
~~~
# The name that we can dlopen(3).
dlname='libmpfr-6.dll'

# Names of this library.
library_names='libmpfr-6.l'

# The name of the static archive.
old_library='libmpfr.lib'
~~~

[1]:https://savannah.gnu.org/patch/?func=detailitem&item_id=9678
[2]:https://gforge.inria.fr/scm/viewvc.php/mpfr/tags/4.0.2/tests/tfits.c?revision=13435&view=co

**Comment by LADSoft** [comment]
I probably won't look at this until tomorrow...

**Comment by LADSoft** [comment]
i tried this but it couldn't find gmp.h...  not sure how to set up MPFR so that it will be found.

**Comment by LADSoft** [comment]
looks like a preprocessor bug, will try to fix tomorrow...

**Comment by GitMensch** [comment]
I've took an existing gmp.dll and header file (from the gnucobol vc dependency download) and created the import library via oimplib. And used LDFLAGS / CPPFLAGS during configure to specify the search directory. Do you need more detailed steps? If yes then I'll redo the steps with a new CI build and adjust the cookbook.

Am 22. Oktober 2019 00:10:41 MESZ schrieb David Lindauer <notifications@github.com>:

**Comment by LADSoft** [comment]
I'm fine at this point, after a lot of hair pulling I got it to build.   Like you say lots of rerunning the configure script and then adjusting libtool while trying to get all the details right lol...    At some point I had to make a side trip to build GMP with mingw so I'd have the right version of the GMP dll. 

I am at the point where I could do make check and have it actually do something;  after tfits.c failed I got the preprocessor output for it.   At first glance it looked like stringizing somehow is broken...
